feat: finish gloss import with lambda

arrocke opened the pull request:
<!--
  Thank you for your PR! Please follow this template to help us review your PR quickly.
-->

## What has changed

* new database record for tracking import jobs
* import lambda and api request to start import are connected by an SQS queue.
* frontend page shows different UI based on whether the import is running, idle or complete.
* Everything is configured in AWS, and I've tested that the lambda code runs when something is put in the queue

<!--
  Please name your PR following conventional commits.
  https://www.conventionalcommits.org/en/v1.0.0/
  The main ones to use are `feat`, `fix`, `chore`, `build`, `refactor`, and `docs`.

  If your PR is not connected to an issue, please describe the reason for your
  changes. Please also describe what you have changed. For more complex changes, include a self review that goes into more detail. Please also call attention to breaking changes, particularly in the shape of API routes.
-->

## Connected Issues

closes #134 

<!--
  If this PR resolves an issue, please add `closes #issue-no` below to connect the issue to the PR.
-->

## QA Steps

1. Comment out the SQS queue code in the post import request. This will allow you to test the lambda locally without SQS.
2. Import the spanish glosses to a language
3. Run `nx serve lambda-functions`
4. When the lambda completes, verify that the import view shows success screen

<!-- Please describe how to test what you've changed. -->

## Post-Deployment

<!--
  Please describe any tasks that must be executed after this change is deployed.
-->

- Upload the latest lambda code to AWS

arrocke committed:
build: lambda working with sqs

arrocke committed:
feat: trigger lambda from api route

arrocke committed:
feat: track language import job status

arrocke committed:
feat: poll the api for import job status

arrocke committed:
refactor: better import notifications

arrocke committed:
refactor: translate interface text in import view

arrocke commented on packages/api-client/src/client.ts:
```diff
@@ -94,25 +94,34 @@ export default class ApiClient {
+    const response = await this.request({ ...request, method: 'POST' });
+    return {
+      ...response.body,
+      location: response.headers.get('Location'),
```
I modified the post request so that we can read the location header to get the id for created resources. Turns out I didn't actually need it, but should be useful in the future.

arrocke commented on packages/db/src/schema.prisma:
```diff
@@ -194,6 +194,22 @@ model Language {
+}
+
+/// Keeps track of a language import job.
+model LanguageImportJob {
```
This new table keeps track of the state of an import job

arrocke commented on packages/api/pages/api/languages/[code]/import.ts:
```diff
@@ -29,23 +61,45 @@ export default createRoute()
-    })),
-    async handler(req, res) {
-      // TODO: query DB for import status
+      if (currentJob && !currentJob.endDate) {
+        return res.conflict([
+          {
+            code: 'JobAlreadyRunning',
+          },
+        ]);
+      }
+
+      await client.languageImportJob.upsert({
```
The import request creates a new job if one isn't already running

Pertempto replied:
Just making sure I'm following: it's an upsert instead of insert because there could be a previously finished job for this language?

arrocke commented on packages/api/pages/api/languages/[code]/import.ts:
```diff
@@ -29,23 +61,45 @@ export default createRoute()
+        },
+      });
+
+      await sqsClient.send(
+        new SendMessageCommand({
+          QueueUrl: process.env.LANGUAGE_IMPORT_QUEUE_URL,
+          MessageGroupId: req.query.code,
+          MessageBody: JSON.stringify({
+            languageCode: req.query.code,
+            importLanguage: req.body.import,
+          }),
+        })
+      );
```
Then it pushes an item on the SQS queue for the lambda to read

arrocke committed:
chore: remove jobid from sqs test event

arrocke commented on packages/lambda-functions/event.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "Records": [
+    {
+      "body": "{ \"jobId\": \"018a5441-617f-ddea-3d5c-a20c4283d24d\", \"languageCode\": \"es\", \"importLanguage\": \"Spanish\" }"
```
This event is used to test the lamdba locally. It mimics the kind of event the lamda would receive from SQS

arrocke commented on packages/lambda-functions/functions/import-language/index.ts:
```diff
@@ -0,0 +1,149 @@
+        }
+      }
+
+      await client.languageImportJob.update({
```
When the import is complete, the lambda updates the job in the database

arrocke commented on packages/web/src/features/languages/ImportLanguageGlossesView.tsx:
```diff
@@ -32,29 +25,89 @@ export async function importLanguageGlossesLoader({
 
   const formContext = useForm<FormData>();
+
+  const [importStatus, setImportStatus] = useState<
+    'idle' | 'running' | 'complete' | 'error'
+  >(() => {
+    if (currentJob) {
+      if (currentJob.endDate) {
+        if (
+          Date.now() - new Date(currentJob.endDate).valueOf() <
```
The import view now looks up the current job to show the running or complete state when you refresh the page.

arrocke commented on packages/web/src/features/languages/ImportLanguageGlossesView.tsx:
```diff
@@ -32,29 +25,89 @@ export async function importLanguageGlossesLoader({
+    }
+  });
+  const pollTimeout = useRef<NodeJS.Timer>();
+  useEffect(() => {
+    if (importStatus === 'running') {
+      const checkImportStatus = async () => {
+        const job = await apiClient.languages.getImportStatus(
+          language.data.code
+        );
```
When the job is running, the frontend polls the backend for the job status on a five second interval.

arrocke committed:
docs: update language import docs

Pertempto commented:
Nice! I'll try to review Tuesday morning

Pertempto commented on packages/api-client/src/auth.ts:
```diff
@@ -15,15 +15,15 @@ export default class Auth {
     });
   }
 
-  login({ email, password }: PostLoginRequest): Promise<void> {
-    return this.client.post({
```
Does this really make any difference? You just don't want to return anything?

Pertempto replied:
You have a lot of changes of this type in this PR. Is it a general cleanup to remove unused return values?

arrocke replied:
It's related to [this comment](https://github.com/arrocke/gloss-translation/pull/176#discussion_r1313956947). Before post requests were returning essentially an empty body, but now they have the location header. For these post requests, the location header isn't relevant so I'm preventing it from being returned from the function.

Pertempto commented on packages/lambda-functions/template.yaml:
```diff
@@ -3,40 +3,29 @@ Transform: AWS::Serverless-2016-10-31
 Globals:
   Function:
-    Timeout: 120
+    Timeout: 360
```
Does this actually work?

arrocke replied:
Yep. This was necessary to prevent the lambda from timing out in local testing

Pertempto commented on packages/web/src/features/languages/ImportLanguageGlossesView.tsx:
```diff
@@ -32,29 +25,89 @@ export async function importLanguageGlossesLoader({
+      if (currentJob.endDate) {
+        if (
+          Date.now() - new Date(currentJob.endDate).valueOf() <
+          10 * 60 * 1000 // 10 minutes
```
Why check for 10 minutes? Why not use the `succeeded` field?

arrocke replied:
If you only use the succeeded flag, then all subsequent views of this page will show the success step. This assumes that if you revisit the page in ten minutes, that you are wanting to start a new import. Maybe that should be 24 hours. It should though determine the status within that 10 minutes based on the succeeded field. I'll fix that

Pertempto reviewed (commented):
The code is looking pretty good! I just had a few questions to help me understand things better. In those cases, it might be good to add more comments to the code since future devs might have the same questions I have.



When I tried to run the lambda function, I get this error:
```
Can't reach database server at `localhost`:`5432`
Please make sure your database server is running at `localhost`:`5432`.
```

What should my `lambda-functions/env.json` file look like? The `env-example.json` still references the old `HelloWorldFunction` name.

arrocke committed:
fix: initial import status for finished jobs

arrocke committed:
fix: env example for lambda functions

arrocke committed:
docs: 24 limit on import view

arrocke commented:
> What should my lambda-functions/env.json file look like? The env-example.json still references the old HelloWorldFunction name

I changed the name of the function to `ImportLanguageFunction` so that needs to be set in the `env.json` file. I updated the example to match.

This should be good for a review again.

Pertempto reviewed (approved):
It seems to work well! In the future, I think a progress indicator would be nice to show the user which book the lambda function is currently working on. This way they don't become worried that the import has stalled out. It wouldn't be too hard to implement, just another field on the job record and in the polling response.
